File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/FlutterAdmobNativeAdsPlugin.kt
package com.tqc.ads.flutter_admob_native_ads

import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.nativead.NativeAd
import com.tqc.ads.flutter_admob_native_ads.ad_loader.NativeAdLoader
import com.tqc.ads.flutter_admob_native_ads.banner.BannerAdLoader
import com.tqc.ads.flutter_admob_native_ads.banner.BannerAdSizeExtensions
import com.tqc.ads.flutter_admob_native_ads.banner.BannerAdViewFactory
import com.tqc.ads.flutter_admob_native_ads.platform_view.NativeAdViewFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * FlutterAdmobNativeAdsPlugin
 *
 * Main plugin class that registers platform views and handles method calls
 * from Flutter for native ad management.
 */
class FlutterAdmobNativeAdsPlugin : FlutterPlugin, MethodCallHandler {

    companion object {
        private const val TAG = "FlutterAdmobNativeAds"
        private const val CHANNEL_NAME = "flutter_admob_native_ads"
        private const val BANNER_CHANNEL_NAME = "flutter_admob_banner_ads"

        // View type identifiers for all 12 forms
        private const val VIEW_TYPE_FORM_1 = "flutter_admob_native_ads_form1"
        private const val VIEW_TYPE_FORM_2 = "flutter_admob_native_ads_form2"
        private const val VIEW_TYPE_FORM_3 = "flutter_admob_native_ads_form3"
        private const val VIEW_TYPE_FORM_4 = "flutter_admob_native_ads_form4"
        private const val VIEW_TYPE_FORM_5 = "flutter_admob_native_ads_form5"
        private const val VIEW_TYPE_FORM_6 = "flutter_admob_native_ads_form6"
        private const val VIEW_TYPE_FORM_7 = "flutter_admob_native_ads_form7"
        private const val VIEW_TYPE_FORM_8 = "flutter_admob_native_ads_form8"
        private const val VIEW_TYPE_FORM_9 = "flutter_admob_native_ads_form9"
        private const val VIEW_TYPE_FORM_10 = "flutter_admob_native_ads_form10"
        private const val VIEW_TYPE_FORM_11 = "flutter_admob_native_ads_form11"
        private const val VIEW_TYPE_FORM_12 = "flutter_admob_native_ads_form12"

        // View type for banner ads
        private const val VIEW_TYPE_BANNER = "flutter_admob_banner_ads"

        // Singleton instance for accessing preloaded ads from platform views
        @Volatile
        private var instance: FlutterAdmobNativeAdsPlugin? = null

        // Static maps to persist across plugin instances (handle multiple onAttachedToEngine calls)
        private val adLoaders = mutableMapOf<String, NativeAdLoader>()
        private val adLoadedCallbacks = mutableMapOf<String, MutableList<(NativeAd) -> Unit>>()
        private val bannerAdLoaders = mutableMapOf<String, BannerAdLoader>()
        private val bannerAdCallbacks = mutableMapOf<String, (AdView) -> Unit>()
        private val loadedBannerAds = mutableMapOf<String, AdView>()

        fun getInstance(): FlutterAdmobNativeAdsPlugin? = instance

        fun getAdLoaders(): Map<String, NativeAdLoader> = adLoaders
        fun setAdLoader(controllerId: String, loader: NativeAdLoader) {
            adLoaders[controllerId] = loader
        }
        fun removeAdLoader(controllerId: String) {
            adLoaders.remove(controllerId)
        }
        fun getAdLoader(controllerId: String): NativeAdLoader? = adLoaders[controllerId]

        fun getBannerAdLoaders(): Map<String, BannerAdLoader> = bannerAdLoaders
        fun setBannerAdLoader(controllerId: String, loader: BannerAdLoader) {
            bannerAdLoaders[controllerId] = loader
        }
        fun removeBannerAdLoader(controllerId: String) {
            bannerAdLoaders.remove(controllerId)
        }
        fun getLoadedBannerAd(controllerId: String): AdView? = loadedBannerAds[controllerId]
        fun setLoadedBannerAd(controllerId: String, adView: AdView) {
            loadedBannerAds[controllerId] = adView
        }

        fun registerNativeAdCallback(controllerId: String, callback: (NativeAd) -> Unit) {
            val callbacks = adLoadedCallbacks.getOrPut(controllerId) { mutableListOf() }
            callbacks.add(callback)
        }

        fun getNativeAdCallbacks(controllerId: String): MutableList<(NativeAd) -> Unit>? {
            return adLoadedCallbacks[controllerId]
        }

        fun clearNativeAdCallbacks(controllerId: String) {
            adLoadedCallbacks.remove(controllerId)
        }

        fun invokeNativeAdCallbacks(controllerId: String, nativeAd: NativeAd) {
            adLoadedCallbacks[controllerId]?.forEach { callback ->
                try {
                    callback(nativeAd)
                } catch (e: Exception) {
                    Log.e(TAG, "Error invoking ad loaded callback for controller: $controllerId", e)
                }
            }
        }

        fun registerBannerAdCallback(controllerId: String, callback: (AdView) -> Unit) {
            bannerAdCallbacks[controllerId] = callback
        }

        fun invokeBannerAdCallback(controllerId: String, adView: AdView) {
            bannerAdCallbacks[controllerId]?.invoke(adView)
        }

        fun clearBannerAdCallback(controllerId: String) {
            bannerAdCallbacks.remove(controllerId)
        }

        fun clearAllAdLoaders() {
            adLoaders.values.forEach { it.destroy() }
            adLoaders.clear()
        }

        fun clearAllBannerAdLoaders() {
            bannerAdLoaders.values.forEach { it.destroy() }
            bannerAdLoaders.clear()
            loadedBannerAds.clear()
        }
    }

    private lateinit var channel: MethodChannel
    private lateinit var bannerChannel: MethodChannel
    private lateinit var context: Context
    private lateinit var messenger: BinaryMessenger

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "Plugin attached to engine")

        instance = this

        context = flutterPluginBinding.applicationContext
        messenger = flutterPluginBinding.binaryMessenger

        // Setup method channel for native ads
        channel = MethodChannel(messenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)

        // Setup method channel for banner ads
        bannerChannel = MethodChannel(messenger, BANNER_CHANNEL_NAME)
        bannerChannel.setMethodCallHandler(this)

        // Register platform view factories
        registerPlatformViews(flutterPluginBinding)
    }

    private fun registerPlatformViews(binding: FlutterPlugin.FlutterPluginBinding) {
        // Register all 12 form layout factories
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_1,
            NativeAdViewFactory(messenger, "form1")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_2,
            NativeAdViewFactory(messenger, "form2")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_3,
            NativeAdViewFactory(messenger, "form3")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_4,
            NativeAdViewFactory(messenger, "form4")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_5,
            NativeAdViewFactory(messenger, "form5")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_6,
            NativeAdViewFactory(messenger, "form6")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_7,
            NativeAdViewFactory(messenger, "form7")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_8,
            NativeAdViewFactory(messenger, "form8")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_9,
            NativeAdViewFactory(messenger, "form9")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_10,
            NativeAdViewFactory(messenger, "form10")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_11,
            NativeAdViewFactory(messenger, "form11")
        )
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_FORM_12,
            NativeAdViewFactory(messenger, "form12")
        )

        // Register banner ad view factory
        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE_BANNER,
            BannerAdViewFactory(messenger)
        )

        Log.d(TAG, "Platform view factories registered: Form1-Form12, Banner")
    }

    /**
     * Gets the preloaded native ad for the given controller ID.
     * Returns null if no ad is loaded for the controller.
     */
    fun getPreloadedAd(controllerId: String): NativeAd? {
        val loader = getAdLoader(controllerId)
        val ad = loader?.getNativeAd()

        Log.d(TAG, "getPreloadedAd($controllerId): loader exists=${loader != null}, ad=$ad")

        return ad
    }

    /**
     * Registers a callback to be invoked when an ad is loaded for the given controller.
     * This allows platform views to receive ads without creating their own loaders.
     *
     * Supports multiple callbacks for the same controllerId (e.g., multiple widgets
     * sharing the same controller).
     */
    fun registerAdLoadedCallback(controllerId: String, callback: (NativeAd) -> Unit) {
        registerNativeAdCallback(controllerId, callback)

        Log.d(TAG, "Registered callback for controller: $controllerId")

        // If ad is already loaded, invoke callback immediately
        getPreloadedAd(controllerId)?.let { ad ->
            callback(ad)
        }
    }

    /**
     * Unregisters the ad loaded callback for the given controller.
     */
    fun unregisterAdLoadedCallback(controllerId: String) {
        clearNativeAdCallbacks(controllerId)
    }

    /**
     * Gets the preloaded banner ad view for the given controller ID.
     * Returns null if no ad is loaded for the controller.
     * Checks the cache first, then falls back to the loader.
     */
    fun getBannerAd(controllerId: String): AdView? {
        val fromCache = getLoadedBannerAd(controllerId)
        val fromLoader = getBannerAdLoaders()[controllerId]?.getAdView()

        Log.d(TAG, "getBannerAd($controllerId): fromCache=$fromCache, fromLoader=$fromLoader")

        return fromCache ?: fromLoader
    }

    /**
     * Registers a callback to be invoked when a banner ad is loaded for the given controller.
     * This allows platform views to receive ads without creating their own loaders.
     */
    fun registerBannerAdCallback(controllerId: String, callback: (AdView) -> Unit) {
        // Call companion object method
        Companion.registerBannerAdCallback(controllerId, callback)

        // If ad is already loaded, invoke callback immediately
        getBannerAd(controllerId)?.let { adView ->
            callback(adView)
        }
    }

    /**
     * Unregisters the banner ad loaded callback for the given controller.
     */
    fun unregisterBannerAdCallback(controllerId: String) {
        clearBannerAdCallback(controllerId)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadAd" -> handleLoadAd(call, result)
            "reloadAd" -> handleReloadAd(call, result)
            "disposeAd" -> handleDisposeAd(call, result)
            "loadBannerAd" -> handleLoadBannerAd(call, result)
            "reloadBannerAd" -> handleReloadBannerAd(call, result)
            "disposeBannerAd" -> handleDisposeBannerAd(call, result)
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            else -> result.notImplemented()
        }
    }

    private fun handleLoadAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")
        val adUnitId = call.argument<String>("adUnitId")
        val enableDebugLogs = call.argument<Boolean>("enableDebugLogs") ?: false

        if (controllerId.isNullOrEmpty() || adUnitId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId and adUnitId are required", null)
            return
        }

        @Suppress("UNCHECKED_CAST")
        val testDeviceIds = call.argument<List<String>>("testDeviceIds")

        // Create and store the loader
        val loader = NativeAdLoader(
            context = context,
            adUnitId = adUnitId,
            controllerId = controllerId,
            messenger = messenger,
            enableDebugLogs = enableDebugLogs,
            testDeviceIds = testDeviceIds
        )

        // Set callback to notify registered platform views
        loader.setOnAdLoadedCallback { nativeAd ->
            invokeNativeAdCallbacks(controllerId, nativeAd)
        }

        setAdLoader(controllerId, loader)
        loader.loadAd()

        result.success(null)
    }

    private fun handleReloadAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")
        val adUnitId = call.argument<String>("adUnitId")
        val enableDebugLogs = call.argument<Boolean>("enableDebugLogs") ?: false

        if (controllerId.isNullOrEmpty() || adUnitId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId and adUnitId are required", null)
            return
        }

        // Destroy existing loader
        getAdLoader(controllerId)?.destroy()

        @Suppress("UNCHECKED_CAST")
        val testDeviceIds = call.argument<List<String>>("testDeviceIds")

        // Create new loader
        val loader = NativeAdLoader(
            context = context,
            adUnitId = adUnitId,
            controllerId = controllerId,
            messenger = messenger,
            enableDebugLogs = enableDebugLogs,
            testDeviceIds = testDeviceIds
        )

        // Set callback to notify registered platform views
        loader.setOnAdLoadedCallback { nativeAd ->
            invokeNativeAdCallbacks(controllerId, nativeAd)
        }

        setAdLoader(controllerId, loader)
        loader.loadAd()

        result.success(null)
    }

    private fun handleDisposeAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")

        if (controllerId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId is required", null)
            return
        }

        getAdLoader(controllerId)?.destroy()
        removeAdLoader(controllerId)

        result.success(null)
    }

    private fun handleLoadBannerAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")
        val adUnitId = call.argument<String>("adUnitId")
        val sizeIndex = call.argument<Int>("size") ?: 5
        val enableDebugLogs = call.argument<Boolean>("enableDebugLogs") ?: false
        val customHeight = call.argument<Int>("adaptiveBannerHeight")

        if (controllerId.isNullOrEmpty() || adUnitId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId and adUnitId are required", null)
            return
        }

        val adSize = BannerAdSizeExtensions.getAdSize(sizeIndex, context, customHeight)

        val loader = BannerAdLoader(
            context = context,
            adUnitId = adUnitId,
            controllerId = controllerId,
            channel = bannerChannel,
            adSize = adSize,
            enableDebugLogs = enableDebugLogs
        )

        loader.setOnAdLoadedCallback { adView ->
            setLoadedBannerAd(controllerId, adView)
            invokeBannerAdCallback(controllerId, adView)
        }

        setBannerAdLoader(controllerId, loader)
        loader.loadAd()

        result.success(null)
    }

    private fun handleReloadBannerAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")
        val adUnitId = call.argument<String>("adUnitId")
        val sizeIndex = call.argument<Int>("size") ?: 5
        val enableDebugLogs = call.argument<Boolean>("enableDebugLogs") ?: false
        val customHeight = call.argument<Int>("adaptiveBannerHeight")

        if (controllerId.isNullOrEmpty() || adUnitId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId and adUnitId are required", null)
            return
        }

        getBannerAdLoaders()[controllerId]?.destroy()

        val adSize = BannerAdSizeExtensions.getAdSize(sizeIndex, context, customHeight)

        val loader = BannerAdLoader(
            context = context,
            adUnitId = adUnitId,
            controllerId = controllerId,
            channel = bannerChannel,
            adSize = adSize,
            enableDebugLogs = enableDebugLogs
        )

        loader.setOnAdLoadedCallback { adView ->
            setLoadedBannerAd(controllerId, adView)
            invokeBannerAdCallback(controllerId, adView)
        }

        setBannerAdLoader(controllerId, loader)
        loader.loadAd()

        result.success(null)
    }

    private fun handleDisposeBannerAd(call: MethodCall, result: Result) {
        val controllerId = call.argument<String>("controllerId")

        if (controllerId.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "controllerId is required", null)
            return
        }

        getBannerAdLoaders()[controllerId]?.destroy()
        removeBannerAdLoader(controllerId)
        clearBannerAdCallback(controllerId)

        result.success(null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        bannerChannel.setMethodCallHandler(null)

        // Clean up all loaders
        clearAllAdLoaders()
        clearAllBannerAdLoaders()

        instance = null
    }
}

File: ios/Classes/Plugin/FlutterAdmobNativeAdsPlugin.swift
import Flutter
import UIKit
import GoogleMobileAds

/// FlutterAdmobNativeAdsPlugin
///
/// Main plugin class that registers platform views and handles method calls
/// from Flutter for native ad and banner ad management.
public class FlutterAdmobNativeAdsPlugin: NSObject, FlutterPlugin {

    private static let channelName = "flutter_admob_native_ads"
    private static let bannerChannelName = "flutter_admob_banner_ads"

    /// Singleton instance for accessing preloaded ads from platform views
    private static var sharedInstance: FlutterAdmobNativeAdsPlugin?

    /// Gets the shared plugin instance
    public static func shared() -> FlutterAdmobNativeAdsPlugin? {
        return sharedInstance
    }

    // View type identifiers for all 12 forms
    private static let viewTypeForm1 = "flutter_admob_native_ads_form1"
    private static let viewTypeForm2 = "flutter_admob_native_ads_form2"
    private static let viewTypeForm3 = "flutter_admob_native_ads_form3"
    private static let viewTypeForm4 = "flutter_admob_native_ads_form4"
    private static let viewTypeForm5 = "flutter_admob_native_ads_form5"
    private static let viewTypeForm6 = "flutter_admob_native_ads_form6"
    private static let viewTypeForm7 = "flutter_admob_native_ads_form7"
    private static let viewTypeForm8 = "flutter_admob_native_ads_form8"
    private static let viewTypeForm9 = "flutter_admob_native_ads_form9"
    private static let viewTypeForm10 = "flutter_admob_native_ads_form10"
    private static let viewTypeForm11 = "flutter_admob_native_ads_form11"
    private static let viewTypeForm12 = "flutter_admob_native_ads_form12"

    // View type for banner ads
    private static let viewTypeBanner = "flutter_admob_banner_ads"

    private var channel: FlutterMethodChannel?
    private var bannerChannel: FlutterMethodChannel?
    private var adLoaders: [String: NativeAdLoader] = [:]

    /// Registry of ad loaded callbacks by controller ID (for platform views)
    /// Changed to support multiple callbacks per controllerId (Array instead of single callback)
    private var adLoadedCallbacks: [String: [(GADNativeAd) -> Void]] = [:]

    /// Registry of banner ad loaders by controller ID
    private var bannerAdLoaders: [String: BannerAdLoader] = [:]

    /// Registry of banner ad loaded callbacks by controller ID (for platform views)
    private var bannerAdCallbacks: [String: (GADBannerView) -> Void] = [:]

    /// Cache of loaded banner ads (to handle race condition between ad load and platform view creation)
    private var loadedBannerAds: [String: GADBannerView] = [:]

    /// Gets the preloaded native ad for the given controller ID.
    /// Returns nil if no ad is loaded for the controller.
    public func getPreloadedAd(controllerId: String) -> GADNativeAd? {
        return adLoaders[controllerId]?.nativeAd
    }

    /// Registers a callback to be invoked when an ad is loaded for the given controller.
    /// This allows platform views to receive ads without creating their own loaders.
    ///
    /// Supports multiple callbacks for the same controllerId (e.g., multiple widgets
    /// sharing the same controller).
    public func registerAdLoadedCallback(controllerId: String, callback: @escaping (GADNativeAd) -> Void) {
        if adLoadedCallbacks[controllerId] == nil {
            adLoadedCallbacks[controllerId] = []
        }
        adLoadedCallbacks[controllerId]?.append(callback)

        let count = adLoadedCallbacks[controllerId]?.count ?? 0
        print("[FlutterAdmobNativeAds] Registered callback for controller: \(controllerId). Total callbacks: \(count)")

        if count > 1 {
            print("[FlutterAdmobNativeAds] ⚠️ WARNING: Multiple callbacks registered for controller: \(controllerId). " +
                    "This may indicate multiple widgets are sharing the same controller. " +
                    "Each widget should have its own NativeAdController instance.")
        }

        // If ad is already loaded, invoke callback immediately
        if let ad = getPreloadedAd(controllerId: controllerId) {
            DispatchQueue.main.async {
                callback(ad)
            }
        }
    }

    /// Unregisters the ad loaded callback for the given controller.
    public func unregisterAdLoadedCallback(controllerId: String) {
        adLoadedCallbacks.removeValue(forKey: controllerId)
    }

    /// Gets the preloaded banner ad view for the given controller ID.
    /// Returns nil if no ad is loaded for the controller.
    /// Checks the cache first, then falls back to the loader.
    public func getBannerAd(controllerId: String) -> GADBannerView? {
        return loadedBannerAds[controllerId] ?? bannerAdLoaders[controllerId]?.getBannerView()
    }

    /// Registers a callback to be invoked when a banner ad is loaded for the given controller.
    /// This allows platform views to receive ads without creating their own loaders.
    public func registerBannerAdCallback(controllerId: String, callback: @escaping (GADBannerView) -> Void) {
        print("[FlutterAdmobNativeAds] Registering banner callback for controller: \(controllerId)")
        bannerAdCallbacks[controllerId] = callback

        // If ad is already loaded, invoke callback immediately
        if let bannerView = getBannerAd(controllerId: controllerId) {
            print("[FlutterAdmobNativeAds] Banner already loaded for controller: \(controllerId), invoking callback immediately")
            DispatchQueue.main.async {
                callback(bannerView)
            }
        }
    }

    /// Unregisters the banner ad loaded callback for the given controller.
    public func unregisterBannerAdCallback(controllerId: String) {
        bannerAdCallbacks.removeValue(forKey: controllerId)
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )

        let bannerChannel = FlutterMethodChannel(
            name: bannerChannelName,
            binaryMessenger: registrar.messenger()
        )

        let instance = FlutterAdmobNativeAdsPlugin()
        instance.channel = channel
        instance.bannerChannel = bannerChannel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addMethodCallDelegate(instance, channel: bannerChannel)

        // Set singleton instance
        FlutterAdmobNativeAdsPlugin.sharedInstance = instance

        // Register all 12 form platform view factories
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form1"),
            withId: viewTypeForm1
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form2"),
            withId: viewTypeForm2
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form3"),
            withId: viewTypeForm3
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form4"),
            withId: viewTypeForm4
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form5"),
            withId: viewTypeForm5
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form6"),
            withId: viewTypeForm6
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form7"),
            withId: viewTypeForm7
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form8"),
            withId: viewTypeForm8
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form9"),
            withId: viewTypeForm9
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form10"),
            withId: viewTypeForm10
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form11"),
            withId: viewTypeForm11
        )
        registrar.register(
            NativeAdViewFactory(messenger: registrar.messenger(), layoutType: "form12"),
            withId: viewTypeForm12
        )

        // Register banner ad view factory
        registrar.register(
            BannerAdViewFactory(messenger: registrar.messenger()),
            withId: viewTypeBanner
        )

        print("[FlutterAdmobNativeAds] Plugin registered with Form1-Form12 layouts and Banner")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadAd":
            handleLoadAd(call, result: result)
        case "reloadAd":
            handleReloadAd(call, result: result)
        case "disposeAd":
            handleDisposeAd(call, result: result)
        case "loadBannerAd":
            handleLoadBannerAd(call, result: result)
        case "reloadBannerAd":
            handleReloadBannerAd(call, result: result)
        case "disposeBannerAd":
            handleDisposeBannerAd(call, result: result)
        case "getPlatformVersion":
            result("iOS \(UIDevice.current.systemVersion)")
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleLoadAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String,
              let adUnitId = args["adUnitId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId and adUnitId are required",
                details: nil
            ))
            return
        }

        let enableDebugLogs = args["enableDebugLogs"] as? Bool ?? false

        print("[FlutterAdmobNativeAds] Loading ad for controller: \(controllerId)")

        guard let channel = channel else {
            result(FlutterError(
                code: "CHANNEL_ERROR",
                message: "Method channel not available",
                details: nil
            ))
            return
        }

        let loader = NativeAdLoader(
            adUnitId: adUnitId,
            controllerId: controllerId,
            channel: channel,
            enableDebugLogs: enableDebugLogs
        )

        // Set delegate to receive ad loaded events
        loader.delegate = self

        adLoaders[controllerId] = loader
        loader.loadAd()

        result(nil)
    }

    private func handleReloadAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String,
              let adUnitId = args["adUnitId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId and adUnitId are required",
                details: nil
            ))
            return
        }

        let enableDebugLogs = args["enableDebugLogs"] as? Bool ?? false

        print("[FlutterAdmobNativeAds] Reloading ad for controller: \(controllerId)")

        // Destroy existing loader
        adLoaders[controllerId]?.destroy()

        guard let channel = channel else {
            result(FlutterError(
                code: "CHANNEL_ERROR",
                message: "Method channel not available",
                details: nil
            ))
            return
        }

        let loader = NativeAdLoader(
            adUnitId: adUnitId,
            controllerId: controllerId,
            channel: channel,
            enableDebugLogs: enableDebugLogs
        )

        // Set delegate to receive ad loaded events
        loader.delegate = self

        adLoaders[controllerId] = loader
        loader.loadAd()

        result(nil)
    }

    private func handleDisposeAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId is required",
                details: nil
            ))
            return
        }

        print("[FlutterAdmobNativeAds] Disposing ad for controller: \(controllerId)")

        adLoaders[controllerId]?.destroy()
        adLoaders.removeValue(forKey: controllerId)

        result(nil)
    }

    private func handleLoadBannerAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String,
              let adUnitId = args["adUnitId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId and adUnitId are required",
                details: nil
            ))
            return
        }

        let sizeIndex = args["size"] as? Int ?? 5
        let enableDebugLogs = args["enableDebugLogs"] as? Bool ?? false
        let customHeight = args["adaptiveBannerHeight"] as? Int

        print("[FlutterAdmobNativeAds] Loading banner ad for controller: \(controllerId)")

        guard let bannerChannel = bannerChannel else {
            result(FlutterError(
                code: "CHANNEL_ERROR",
                message: "Banner method channel not available",
                details: nil
            ))
            return
        }

        let adSize = BannerAdSizeExtensions.getAdSize(sizeIndex: sizeIndex, customHeight: customHeight)

        let loader = BannerAdLoader(
            adUnitId: adUnitId,
            controllerId: controllerId,
            channel: bannerChannel,
            adSize: adSize,
            enableDebugLogs: enableDebugLogs
        )

        loader.delegate = self

        // Set callback to notify registered platform views
        loader.setOnAdLoadedCallback { [weak self] bannerView in
            guard let self = self else { return }
            print("[FlutterAdmobNativeAds] Banner ad loaded callback invoked for controller: \(controllerId)")
            print("[FlutterAdmobNativeAds] Registered callbacks: \(self.bannerAdCallbacks.keys)")

            // Cache the bannerView to handle race condition with platform view creation
            self.loadedBannerAds[controllerId] = bannerView

            if let callback = self.bannerAdCallbacks[controllerId] {
                print("[FlutterAdmobNativeAds] Invoking callback for controller: \(controllerId)")
                DispatchQueue.main.async {
                    callback(bannerView)
                }
            } else {
                print("[FlutterAdmobNativeAds] WARNING: No callback registered for controller: \(controllerId)")
            }
        }

        bannerAdLoaders[controllerId] = loader
        loader.loadAd()

        result(nil)
    }

    private func handleReloadBannerAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String,
              let adUnitId = args["adUnitId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId and adUnitId are required",
                details: nil
            ))
            return
        }

        let sizeIndex = args["size"] as? Int ?? 5
        let enableDebugLogs = args["enableDebugLogs"] as? Bool ?? false
        let customHeight = args["adaptiveBannerHeight"] as? Int

        print("[FlutterAdmobNativeAds] Reloading banner ad for controller: \(controllerId)")

        // Destroy existing loader
        bannerAdLoaders[controllerId]?.destroy()

        guard let bannerChannel = bannerChannel else {
            result(FlutterError(
                code: "CHANNEL_ERROR",
                message: "Banner method channel not available",
                details: nil
            ))
            return
        }

        let adSize = BannerAdSizeExtensions.getAdSize(sizeIndex: sizeIndex, customHeight: customHeight)

        let loader = BannerAdLoader(
            adUnitId: adUnitId,
            controllerId: controllerId,
            channel: bannerChannel,
            adSize: adSize,
            enableDebugLogs: enableDebugLogs
        )

        loader.delegate = self

        // Set callback to notify registered platform views
        loader.setOnAdLoadedCallback { [weak self] bannerView in
            guard let self = self else { return }
            print("[FlutterAdmobNativeAds] Banner ad loaded callback invoked for controller: \(controllerId)")
            print("[FlutterAdmobNativeAds] Registered callbacks: \(self.bannerAdCallbacks.keys)")

            // Cache the bannerView to handle race condition with platform view creation
            self.loadedBannerAds[controllerId] = bannerView

            if let callback = self.bannerAdCallbacks[controllerId] {
                print("[FlutterAdmobNativeAds] Invoking callback for controller: \(controllerId)")
                DispatchQueue.main.async {
                    callback(bannerView)
                }
            } else {
                print("[FlutterAdmobNativeAds] WARNING: No callback registered for controller: \(controllerId)")
            }
        }

        bannerAdLoaders[controllerId] = loader
        loader.loadAd()

        result(nil)
    }

    private func handleDisposeBannerAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let controllerId = args["controllerId"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "controllerId is required",
                details: nil
            ))
            return
        }

        print("[FlutterAdmobNativeAds] Disposing banner ad for controller: \(controllerId)")

        bannerAdLoaders[controllerId]?.destroy()
        bannerAdLoaders.removeValue(forKey: controllerId)
        bannerAdCallbacks.removeValue(forKey: controllerId)
        loadedBannerAds.removeValue(forKey: controllerId)

        result(nil)
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        print("[FlutterAdmobNativeAds] Plugin detached from engine")

        // Clean up all loaders
        adLoaders.values.forEach { $0.destroy() }
        adLoaders.removeAll()

        // Clean up all banner loaders
        bannerAdLoaders.values.forEach { $0.destroy() }
        bannerAdLoaders.removeAll()
        loadedBannerAds.removeAll()

        // Clear callbacks
        adLoadedCallbacks.removeAll()
        bannerAdCallbacks.removeAll()

        // Clear channels
        channel = nil
        bannerChannel = nil

        // Clear singleton instance
        FlutterAdmobNativeAdsPlugin.sharedInstance = nil
    }
}

// MARK: - NativeAdLoaderDelegate

extension FlutterAdmobNativeAdsPlugin: NativeAdLoaderDelegate {

    func adLoader(_ loader: NativeAdLoader, didReceiveNativeAd nativeAd: GADNativeAd) {
        // Notify ALL registered callbacks for this controller
        if let callbacks = adLoadedCallbacks[loader.controllerId] {
            DispatchQueue.main.async {
                for callback in callbacks {
                    callback(nativeAd)
                }
            }
        }
    }

    func adLoader(_ loader: NativeAdLoader, didFailWithError error: Error) {
        // Error is already sent via method channel by the loader
        print("[FlutterAdmobNativeAds] Ad failed to load: \(error.localizedDescription)")
    }
}

// MARK: - BannerAdLoaderDelegate

extension FlutterAdmobNativeAdsPlugin: BannerAdLoaderDelegate {

    func adLoader(_ loader: BannerAdLoader, didReceiveBannerAd bannerView: GADBannerView) {
        // Notify registered callbacks for this controller
        if let callback = bannerAdCallbacks[loader.controllerId] {
            DispatchQueue.main.async {
                callback(bannerView)
            }
        }
    }

    func adLoader(_ loader: BannerAdLoader, didFailWithError error: Error) {
        // Error is already sent via method channel by the loader
        print("[FlutterAdmobNativeAds] Banner ad failed to load: \(error.localizedDescription)")
    }
}
